File: app/src/main/java/com/ivanmagda/habito/utils/ReminderUtils.java
package com.ivanmagda.habito.utils;

import android.app.AlarmManager;
import android.app.PendingIntent;
import android.content.Context;
import android.content.Intent;
import android.support.annotation.NonNull;

import com.ivanmagda.habito.models.Habit;
import com.ivanmagda.habito.receivers.ReminderReceiver;

import java.util.Calendar;
import java.util.List;

public class ReminderUtils {

    private static final String TAG = "ReminderUtils";

    public static final String HABIT_EXTRA_KEY = "habit";

    private static PendingIntent buildPendingIntent(@NonNull final Habit habit,
                                                    @NonNull final Context context) {
        Intent alarmIntent = new Intent(context, ReminderReceiver.class);
        alarmIntent.putExtra(HABIT_EXTRA_KEY, habit);
        return PendingIntent.getBroadcast(context, (int) habit.getRecord().getCreatedAt(),
                alarmIntent, PendingIntent.FLAG_UPDATE_CURRENT);
    }

    public static void setAlarm(@NonNull final Habit habit, @NonNull final Context context) {
        AlarmManager alarmManager = (AlarmManager) context.getSystemService(Context.ALARM_SERVICE);
        PendingIntent pendingIntent = buildPendingIntent(habit, context);

        if (!habit.isReminderOn()) {
            alarmManager.cancel(pendingIntent);
        } else {
            Calendar now = Calendar.getInstance();
            now.setTimeInMillis(System.currentTimeMillis());
            Calendar alarmTime = Calendar.getInstance();
            alarmTime.setTimeInMillis(System.currentTimeMillis());
            alarmTime.set(Calendar.HOUR_OF_DAY, habit.getRecord().getReminderHour());
            alarmTime.set(Calendar.MINUTE, habit.getRecord().getReminderMin());
            alarmTime.set(Calendar.SECOND, 0);

            if (now.after(alarmTime)) {
                alarmTime.add(Calendar.DATE, 1);
            }

            alarmManager.cancel(pendingIntent);
            alarmManager.setRepeating(AlarmManager.RTC_WAKEUP, alarmTime.getTimeInMillis(),
                    AlarmManager.INTERVAL_DAY, pendingIntent);
        }
    }

    public static void cancelAlarm(@NonNull final Habit habit, @NonNull final Context context) {
        AlarmManager alarmManager = (AlarmManager) context.getSystemService(Context.ALARM_SERVICE);
        PendingIntent pendingIntent = buildPendingIntent(habit, context);
        alarmManager.cancel(pendingIntent);
    }

    public static void processOn(@NonNull final Habit habit, @NonNull final Context context) {
        if (habit.isReminderOn()) {
            setAlarm(habit, context);
        } else {
            cancelAlarm(habit, context);
        }
    }

    public static void processAll(@NonNull final List<Habit> habits, @NonNull final Context context) {
        for (Habit habit : habits) {
            processOn(habit, context);
        }
    }

    public static void scheduleAll(@NonNull final List<Habit> habits, @NonNull final Context context) {
        for (Habit habit : habits) {
            setAlarm(habit, context);
        }
    }

    public static void cancelAll(@NonNull final List<Habit> habits, @NonNull final Context context) {
        for (Habit habit : habits) {
            cancelAlarm(habit, context);
        }
    }

    private ReminderUtils() {
    }

}

File: app/src/main/java/com/ivanmagda/habito/utils/ReminderUtils.kt
package com.ivanmagda.habito.utils

import android.app.AlarmManager
import android.app.PendingIntent
import android.content.Context
import android.content.Intent
import com.ivanmagda.habito.models.Habit
import com.ivanmagda.habito.receivers.ReminderReceiver
import java.util.*

object ReminderUtils {

    val HABIT_EXTRA_KEY = "habit"

    fun processHabit(habit: Habit, context: Context) {
        if (habit.isReminderOn) {
            setAlarm(habit, context)
        } else {
            cancelAlarm(habit, context)
        }
    }

    fun processAll(habits: List<Habit>, context: Context) {
        for (habit in habits) {
            processHabit(habit, context)
        }
    }

    fun scheduleAll(habits: List<Habit>, context: Context) {
        for (habit in habits) {
            setAlarm(habit, context)
        }
    }

    fun cancelAll(habits: List<Habit>, context: Context) {
        for (habit in habits) {
            cancelAlarm(habit, context)
        }
    }

    private fun buildPendingIntent(habit: Habit, context: Context): PendingIntent {
        val alarmIntent = Intent(context, ReminderReceiver::class.java)
        alarmIntent.putExtra(HABIT_EXTRA_KEY, habit)

        return PendingIntent.getBroadcast(context, habit.record.createdAt.toInt(), alarmIntent,
                PendingIntent.FLAG_UPDATE_CURRENT)
    }

    private fun setAlarm(habit: Habit, context: Context) {
        val alarmManager = context.getSystemService(Context.ALARM_SERVICE) as AlarmManager
        val pendingIntent = buildPendingIntent(habit, context)

        if (!habit.isReminderOn) {
            alarmManager.cancel(pendingIntent)
        } else {
            val now = Calendar.getInstance()
            now.timeInMillis = System.currentTimeMillis()

            val alarmTime = Calendar.getInstance()
            alarmTime.timeInMillis = System.currentTimeMillis()
            alarmTime.set(Calendar.HOUR_OF_DAY, habit.record.reminderHour)
            alarmTime.set(Calendar.MINUTE, habit.record.reminderMin)
            alarmTime.set(Calendar.SECOND, 0)

            if (now.after(alarmTime)) {
                alarmTime.add(Calendar.DATE, 1)
            }

            alarmManager.cancel(pendingIntent)
            alarmManager.setRepeating(AlarmManager.RTC_WAKEUP, alarmTime.timeInMillis,
                    AlarmManager.INTERVAL_DAY, pendingIntent)
        }
    }

    private fun cancelAlarm(habit: Habit, context: Context) {
        val alarmManager = context.getSystemService(Context.ALARM_SERVICE) as AlarmManager
        val pendingIntent = buildPendingIntent(habit, context)
        alarmManager.cancel(pendingIntent)
    }
}
